### PR TITLE
Restore BC for the tensor-parallel API

### PR DESCRIPTION
**Pull request by ArthurZucker**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48300) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48300)

Restore `tp_plan` from `from_pretrained()` and add deprecation cycle to it
More guarding regarding `qk_rope_head_dim` attributes in MLA + expert dim floor-divided by mesh size 
Doc updates

### CODE REVIEW

File: src/transformers/distributed/configuration_utils.py [hunk 1]
```diff
@@ -50,6 +50,9 @@ class DistributedConfig:
     pp_size: int | None = None
 
     def __post_init__(self):
+        if self.tp_plan is not None and self.tp_size is None and self.fsdp_size is None and self.pp_size is None:
```
Comment (3outeille) [inline src/transformers/distributed/configuration_utils.py]: there is ` if self.tp_size is None: self.tp_size = 1` just below which can be reuse

File: src/transformers/distributed/mixin.py [hunk 2]
```diff
@@ -73,6 +72,19 @@ def init_parallel_plans(self) -> None:
...
+                    input_embeddings = self.get_input_embeddings()
+                except NotImplementedError:
+                    input_embeddings = None
+                if isinstance(input_embeddings, nn.Embedding):
+                    input_embedding_name = next(
+                        (name for name, module in self.named_modules() if module is input_embeddings), None
+                    )
+                    if input_embedding_name:
+                        self._tp_plan.setdefault(input_embedding_name, "embedding_rowwise")
+
```
Comment (ArthurZucker) [inline src/transformers/distributed/mixin.py]: this is ai slope

File: src/transformers/distributed/mixin.py [hunk 3]
```diff
@@ -120,14 +132,19 @@ def tp_plan(self, plan: dict[str, str] | None):
...
+        def _matches(pattern: str) -> bool:
+            regex_pattern = pattern.replace("*", r"\d+")
+            return any(re.match(regex_pattern, name) for name in model_param_names)
+
+        prefix = getattr(self, "base_model_prefix", "")
+        if prefix:
+            plan = {
+                f"{prefix}.{k}" if not _matches(k) and _matches(f"{prefix}.{k}") else k: v for k, v in plan.items()
+            }
+
```
Comment (ArthurZucker) [inline src/transformers/distributed/mixin.py]: this is also ai slope

File: src/transformers/distributed/tensor_parallel.py [hunk 4]
```diff
@@ -775,6 +788,13 @@ class ParallelInterface(GeneralInterface):
 def apply_tensor_parallelism(model, tp_mesh):
     """DTensor backend: shard params as placeholders and install TP forward hooks."""
 
+    for layer_pattern, style in (model.tp_plan or {}).items():
+        if style not in ALL_PARALLEL_STYLES:
+            raise ValueError(
+                f"Unsupported tensor parallel style '{style}' for layer '{layer_pattern}'. "
+                f"Supported styles are {list(ALL_PARALLEL_STYLES.keys())}"
+            )
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py]: no, collect all errors, then value error with all unrecognized

File: src/transformers/distributed/utils.py [hunk 5]
```diff
@@ -67,6 +70,11 @@ def _ensure_torch_distributed(device_type: str | None = None):
     if not torch.distributed.is_initialized():
         if device_type is None:
             device_type = torch._C._get_accelerator().type
+        if device_type == "mps":
+            logger.warning_once(
+                "PyTorch's built-in DeviceMesh/DTensor stack does not support an MPS mesh. Falling back to CPU."
+            )
```
Comment (ArthurZucker) [inline src/transformers/distributed/utils.py:76]: ok

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -348,12 +348,6 @@ def __post_init__(self, **kwargs):
         if per_layer_config is not None:
             self.per_layer_config = per_layer_config
 
-        if getattr(self, "tie_word_embeddings", False) and self.base_model_tp_plan is not None:
-            self.base_model_tp_plan = {
-                **self.base_model_tp_plan,
-                "embed_tokens": "embedding_rowwise",
```
Comment (ArthurZucker) [inline src/transformers/configuration_utils.py:354]: This was good, if not present and tie, keep it

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -4184,6 +4186,26 @@ def from_pretrained(
...
+                "Pass either `distributed_config` or the standalone `tp_plan`/`tp_size` arguments, not both. "
+                "Set tensor-parallel options on `DistributedConfig` when using it."
+            )
+
+        if tp_plan is not None:
+            warnings.warn(
+                "Passing `tp_plan` directly to `from_pretrained` is deprecated and will be removed in v5.18. "
+                "Pass it in `distributed_config=DistributedConfig(tp_plan=...)` instead.",
+                FutureWarning,
+                stacklevel=2,
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:4203]: nice

File: docs/source/en/model_doc/ernie4_5_vl_moe.md [hunk 8]
```diff
@@ -85,7 +85,7 @@ from transformers import AutoModelForImageTextToText, AutoProcessor
 
 model = AutoModelForImageTextToText.from_pretrained(
     "baidu/ERNIE-4.5-VL-28B-A3B-PT",
-    device_map="auto",  # Use tp_plan="auto" instead to enable Tensor Parallelism!
+    device_map="auto",  # Pass distributed_config=DistributedConfig(tp_size=N) instead to enable TP!
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/ernie4_5_vl_moe.md]: this is not ideal. It's annoying to write. 
Can we either: 
- keep tp plan auto
- auto detect torch run no device map -> tp plan auto?

File: docs/source/en/model_doc/llama4.md [hunk 9]
```diff
@@ -366,14 +368,20 @@ tokenizer = AutoTokenizer.from_pretrained(model_id)
...
     {"role": "user", "content": "Who are you?"},
 ]
-inputs = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt", return_dict=True).to(model.device)
 
+distributed_config = DistributedConfig(tp_size=int(os.environ["WORLD_SIZE"]))
 model = Llama4ForConditionalGeneration.from_pretrained(
     model_id,
-    tp_plan="auto",
-    device_map="auto",
+    distributed_config=distributed_config,
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/llama4.md]: not a good default at all. why force the use to compute world size, import os etc when a single simple tp plan was enough before

File: docs/source/en/model_doc/ernie4_5_vl_moe.md (hunk 8 above)
Comment (3outeille) [reply docs/source/en/model_doc/ernie4_5_vl_moe.md]: yeah but your suggestions dont take into account other parallelism. I agree that having `tp_plan=auto` works wonder but if you want to combine with fsdp or pp, then `DistributedConfig(tp_size=2, fsdp_size=2, pp_size=2)` makes more sense

File: docs/source/en/model_doc/llama4.md (hunk 9 above)
Comment (3outeille) [reply docs/source/en/model_doc/llama4.md]: you can use `DistributedConfig(tp_plan=auto)` if you want but with other parallelism, that would be weird to do`DistributedConfig(tp_plan=auto, fsdp_size=2, pp_size=2)` for example

Comment (ArthurZucker) [reply docs/source/en/model_doc/llama4.md]: absolutely, but we need to remove friction for the more common use case : inference!

File: docs/source/en/model_doc/ernie4_5_vl_moe.md (hunk 8 above)
Comment (ArthurZucker) [reply docs/source/en/model_doc/ernie4_5_vl_moe.md]: what I am saying and what you are saying are not against one another! its more about good defaults! DP is never used in inference, PP can be yes -> we would ideally want the torch run command to be caught and auto infer tp size and pp size (inter / intra)

Comment (3outeille) [reply docs/source/en/model_doc/ernie4_5_vl_moe.md]: okay let's say we go for `DistributedConfig(tp_plan=auto, pp_plan=auto)` and `WORLD_SIZE=16`, 

we would intuitively infer `tp_size=8` and `pp_size=2` but if some reasons, someone would want to do `tp_size=4` and `pp_size=4`, we loose the ability to do so since we auto infer.

On top of that, it's heavily dependant on the infra but sometimes PP can be faster than TP. Here some benchmark Ive ran back in the days (Same DP, batch settings, nodes, and GPU count. PP is 10.8% faster) 

PP: dp16_tp1_pp4_mbs1_ga64  — 5211 tok/s/GPU  (source: https://huggingface.co/datasets/nanotron/picotron_bench/tree/main/result_blog_posts/saturate_dp_7b_8_node_pp/dp16_tp1_pp4_mbs1_ga64_sl4096)
TP: dp16_tp4_pp1_mbs1_ga64  — 4705 tok/s/GPU (source: https://huggingface.co/datasets/nanotron/picotron_bench/tree/main/result_blog_posts/saturate_dp_7b_8_node_tp/dp16_tp4_pp1_mbs1_ga64_sl4096)
